### PR TITLE
C++ 11 support: replace auto_ptr with unique_ptr

### PR DESCRIPTION
**Pull request by D4N**
This is an enormous PR, consisting of mostly enormous commits which are mass replaces of `auto_ptr` with `unique_ptr`.

### CODE REVIEW

**Comment by piponazo** [comment]
Have you discussed this change with Robin? I remember we had a discussion few months ago about switching to c++11 and we decided to do it for a later release.

**Comment by D4N** [comment]
Yes, there is an issue for C++11 support with all the details: #214

**Comment by D4N** [comment]
@piponazo Do you know if the Appveyor build can be restarted? The conan server seemed to be down, thus failing the build.

**Comment by piponazo** [comment]
Good, now the CI check for appveyor passed 👍 

I'm very happy to see that we will start to use c++11. Right now I'm in a short holidays. I'll try to give a quick review to the PR and in Monday/Tuesday continue with it.

File: config/compilerFlags.cmake
Comment (piponazo) [inline config/compilerFlags.cmake]: The best way to require C++11 with CMake is by setting the following global CMake variables:

```
set(CMAKE_CXX_STANDARD 11)
set(CXX_STANDARD_REQUIRED ON)
set(CMAKE_CXX_EXTENSIONS OFF)
```
This will translate to the specific compiler flags for different compilers. I suggest to remove the `-std=c++11` flag here and use those variables.

**Review by piponazo — CHANGES_REQUESTED** [review]
All the changes in the c++ code look good to me. I just would like to ask you to change the way in which we specify the c++11 requirement in the CMake code.

Then I have a question. Do we want to continue supporting the autotools code? I did not see anything about that in #214, and here we are not updating it.

File: config/compilerFlags.cmake
Comment (D4N) [reply config/compilerFlags.cmake]: Didn't know that. I'll include that, that is a lot better than hard coding it via `-std=c++11`. Thanks for the tip!

**Comment by D4N** [comment]
Uh, to be honest, I completely forgot about autotools... I can add the flags to that, if I find them. @clanmills can you take a look at the result then? I know basically nothing about autotools.

**Comment by clanmills** [comment]
Sure, I'll do that.  I don't want you to spend a minute on autotools.  Your skills are so much better deployed everywhere else!   I think I added c++11/14/17 last year.  I have a sniff into this tomorrow.

**Comment by D4N** [comment]
@piponazo Done.
@clanmills You have indeed added the option to build with C++11, I just changed the default to C++11, otherwise the default build would fail now.

**Review by piponazo — APPROVED** [review]

**Comment by clanmills** [comment]
Gentlemen:

I did add support for C++11 (and 14 and 17) to the autotools build environment in 2017 when working on Adobe XMPsdk.  I'm pleased that 'master' builds and passes the test suite on when I build with C++11 

```
504 rmills@rmillsmbp:~/gnu/github/exiv2/exiv2 $ ./configure --help | grep -e std -e adobe
  --with-std  use --with-std for c++11. options: --with-std=11|14|c++11|c++14 default/--without-std is c++98
  --with-adobe[=OPT]  use --with-adobe for Adobe XMPsdk. OPT: 2016|2014|2013 default/--without-adobe is xmpsdk/src
505 rmills@rmillsmbp:~/gnu/github/exiv2/exiv2 $
```

However, I believe I'm building with the auto-ptr and more work is necessary to use unique-ptr.  In a conversation with Dan and Michal, I was lead to believe that there was fundemental difficulty in getting that to work (something to with copying a pointer).  

I think we should proceed with some caution here and discuss this carefully together in England in May.  We have lots of work-in-progress leading up to v0.27.  I would like to offer an API that can be called by C++98 applications (and prefer the library to be build with C++98).  In the release notes for v0.27, we can state our future intention to deprecate C++98 support.

I believe the issue with the unique-ptr arises in the TiffMumble code.  I would like us to have a discussion about that code as it may be considerably refactored to support BigTiff (64 bit tiff).

My feeling is that we should release v0.27 with all the security fixes and retain C++98 support and many editions of Visual Studio.  The "modernisation" of the code should be announced as the "key-note" feature of v0.28 (which may even be called v1.0).

**Comment by D4N** [comment]
The issues with moving from `auto_ptr` to `unique_ptr` were actually trivial to solve, it amounted to explicitly calling `std::move()` here and there. The "problem" with `auto_ptr` is its weird copy-behavior, i.e. if you call:
``` c++
std::auto_ptr ptr_b = ptr_a;
```
then `ptr_a`'s pointer is moved to `ptr_b` and `ptr_a` is then `NULL`. `unique_ptr` does exactly the same, only you have to be explicit about the move (a copy won't compile):
``` c++
std::unique_ptr ptr_b = std::move(ptr_a);
```
All the necessary work to get `unique_ptr` into the code base has been addressed by this PR. I was afraid it would be more work, but `auto_ptr` was used only in places where you can perform a simple search & replace without any further problems.

If we really want to be C++98 compatible, then we cannot merge this PR, as it will break C++98 compatibility and require C++>11. However, I though we wanted to switch to C++11 for 0.27?

**Comment by D4N** [comment]
To be honest, I think we should integrate this for 0.27 already, mainly for the sake of getting this update in a reasonable amount of time downstream.

Consider this: it's already 2018 and it's going to take a while until 0.27 is released, so we can expect 0.28 sometime 2019ish? Add to that the packaging delay of a few months to a year and we can easily end up in 2020 before downstream gets the version of exiv2 without `auto_ptr`. At that point, `auto_ptr` has been **removed** from the most recent C++ standard for 3 years, which could lead to significant frustration of downstream users. It may very well be too late to avoid that, since C++17 is already a thing.

**Comment by clanmills** [comment]
I'm OK about removing this. Remember that the API and the implementation can be different.  It's possible for the implementation to be in C++17 and offer and API that can be used by C++98.

My concern about legacy is to retain an API/library that can be used by older compilers/build systems.  I was very surprised when contracting in Silicon Valley to discover MSVC v6.0 (1998) in use at more than one company (not Adobe) who had legacy code that they occasionally need to modify/rebuild and didn't want to update the tools/build-chain.  Adobe were the opposite.  They would upgrade something to the latest Xcode/MSVC and that was that.  They didn't care if the code didn't build with older tools.  In open-source, we should try to satisfy both worlds by building with as many build-chains as possible.

**Comment by piponazo** [comment]
I totally agree with @D4N. Apart for the reasons he gave (that are the important ones), it will make our work much more pleasant, giving us the opportunity of collaborating in an important Open Source project while using a recent language standard. 

At work and in my personal projects I already use (at least) C++11. When I have to collaborate in Exiv2 and change back to C++98 I feel a bit frustrated. 

If we decide to remove any usage of `auto_ptr` from the public headers, then it will take more work. Did any of you analyze that possibility already? I could take a look at that to analyze if it is feasible.

**Comment by D4N** [comment]
On April 7, 2018 8:02:21 AM GMT+02:00, "Luis Díaz Más" <notifications@github.com> wrote: 
>
>If we decide to remove any usage of `auto_ptr` from the public headers,
>then it will take more work. Did any of you analyze that possibility
>already? I could take a look at that to analyze if it is feasible.

No need to do that, that's exactly what this PR does: it completely purges auto_ptr from the codebase, especially from the public headers.

**Comment by piponazo** [comment]
@D4N I meant to remove `auto_ptr` from the public headers so that we do not use `std::unique_ptr` either there after the switch to C++11. 

I just started to analyze the topic for one particular case and I think it is possible to do it. But I can continue the analysis once this is merged, and try to clean up the usages of `unique_ptr` where it is not needed. Now that we agree on this PR, let's merge it! :confetti_ball:

**Comment by piponazo** [comment]
Fine for me @D4N , I think it is what we concluded.

The error in CI says this:

```
[ 57%] Building CXX object unitTests/CMakeFiles/unit_tests.dir/test_XmpKey.cpp.o
/home/travis/build/Exiv2/exiv2/unitTests/test_XmpKey.cpp: In member function ‘virtual void AXmpKey_canBeCloned_Test::TestBody()’:
/home/travis/build/Exiv2/exiv2/unitTests/test_XmpKey.cpp:69:5: error: ‘AutoPtr’ is not a member of ‘Exiv2::XmpKey’
     XmpKey::AutoPtr clonedKey = key.clone();
```
It seems there is a AutoPtr leftover.

**Comment by D4N** [comment]
Yes, I saw the appveyor build fail & took a look at it when I was typing the last comment. I think the `AutoPtr` somehow slipped through when I rebased my changes on master.

**Comment by piponazo** [comment]
Ey @D4N. Do you mind to close this PR until we release 0.27. You can always keep the branch in your fork, and fix the conflicts once you resume the work.

I think it would be good to have as less PRs opened as possible, so that we can focus in the things going on.

**Comment by D4N** [comment]
Defered until 0.28

**Pull request closed without merging**
